libs/pika/async_mpi/include/pika/async_mpi/dispatch_mpi.hpp

```diff
@@ -0,0 +1,268 @@
+                                static_assert(std::is_same_v<invoke_result_type, int>);
+                                status = PIKA_INVOKE(PIKA_MOVE(r.op_state.f), ts..., &request);
+                            }
+                            PIKA_DETAIL_DP(mpi_tran<7>,
+                                debug(str<>("dispatch_mpi_recv"), "invoke mpi",
+                                    detail::stream_name(r.op_state.stream_), ptr(request)));
+
+                            PIKA_ASSERT_MSG(request != MPI_REQUEST_NULL,
+                                "MPI_REQUEST_NULL returned from mpi invocation");
+
+                            if (status != MPI_SUCCESS)
+                            {
+                                PIKA_DETAIL_DP(mpi_tran<5>,
+                                    debug(str<>("set_error"), "status != MPI_SUCCESS",
+                                        detail::error_message(status)));
+                                ex::set_error(PIKA_MOVE(r.op_state.receiver),
+                                    std::make_exception_ptr(mpi_exception(status)));
+                                return;
+                            }
+
+                            if (poll_request(request))
+                            {
+                                PIKA_DETAIL_DP(mpi_tran<7>,
+                                    debug(str<>("dispatch_mpi_recv"), "eager poll ok",
+                                        detail::stream_name(r.op_state.stream_), ptr(request)));
+                                // calls set_value(request), or set_error(mpi_exception(status))
+                                set_value_error_helper(
+                                    status, PIKA_MOVE(r.op_state.receiver), MPI_REQUEST_NULL);
+                            }
+                            else
+                            {
+                                set_value_error_helper(
+                                    status, PIKA_MOVE(r.op_state.receiver), request);
+                            }
+                        },
+                        [&](std::exception_ptr ep) {
+                            ex::set_error(PIKA_MOVE(r.op_state.receiver), PIKA_MOVE(ep));
+                        });
+                }
+
+                friend constexpr ex::empty_env tag_invoke(
+                    ex::get_env_t, dispatch_mpi_receiver const&) noexcept
+                {
+                    return {};
+                }
+            };
+
+            using operation_state_type =
+                ex::connect_result_t<std::decay_t<Sender>, dispatch_mpi_receiver>;
+            operation_state_type op_state;
+
+            template <typename Tuple>
+            struct value_types_helper
+            {
+                using type = pud::transform_t<Tuple, std::decay>;
+            };
+
+            // we don't bother storing the result, but the correct type is ...
+            using result_type = pika::detail::variant<std::tuple<MPI_Request>>;
```

Comment:
If we don't bother storing the result, why leave it here?
```suggestion
```

Reply:
because it takes me ges to find the right type when I change anything and so having it there is a reminder to my future self when I decide to modify the result for some reason. I will remove it